File: Algorithm/inflearn_cpp/myCode/section5_DP/네트워크_선_자르기/네트워크선자르기_top_down.cpp
#include <iostream>
using namespace std;

int dy[101];

int dfs(int n)
{
  if (dy[n] != 0)
    return dy[n];
  if (n == 1 || n == 2)
    return n;
  return dy[n] = dfs(n - 1) + dfs(n - 2);
}

int main()
{
  ios_base::sync_with_stdio(false);
  freopen("input.txt", "rt", stdin);
  int n;
  cin >> n;
  cout << dfs(n) << endl;
  return 0;
}

File: Algorithm/inflearn_cpp/myCode/section5_DP/네트워크_선_자르기/네트워크선자르기_top_down.py
import sys
# sys.stdin = open('input.txt')
dy = [0] * 50


def dfs(n):
    if dy[n]:
        return dy[n]
    if n == 1 or n == 2:
        return n
    dy[n] = dfs(n-1) + dfs(n-2)
    return dy[n]

length = int(input())


print(dfs(length))
